editor-server/src/routes/api/uploadData.ts

```diff
@@ -42,8 +70,51 @@ const uploadData = async (req: Request, res: Response) => {
     await prisma.dancer.deleteMany();
     await prisma.positionFrame.deleteMany();
     await prisma.controlFrame.deleteMany();
+    await prisma.lEDFrame.deleteMany();
+    await prisma.lEDEffect.deleteMany();
+
+    // check all data are valid
+
+    // add fiber color list
+    Object.keys(color).map((colorKey: string) => {
+      allFiberColors.add(colorKey);
+    });
+    // add LED part list
+    dancer.map((dancerObj: TDancerData) => {
+      const { parts, name } = dancerObj;
+      // console.log(allPartsList)
+      allPartsList[name] = parts.map((partObj: TPartData) => {
+        const { name, type } = partObj;
+        if (type === "LED") allLEDPartNames.add(name);
+        return partObj;
+      });
+    });
+    // check LED part is used by any dancer
+    Object.keys(ledEffects).map((partName: string) => {
+      if (!allLEDPartNames.has(partName)) {
+        throw new Error(`LED part '${partName}' is not used by any dancer!!`);
+      }
+    });
+    const sortedDancerTmp = Object.keys(allPartsList).sort();
+    Object.values(control).map((frameObj: CtrlFrameTmpData) => {
+      const { fade, start, status } = frameObj;
+
+      for (let i = 0; i < status.length; i++) {
+        for (let j = 0; j < status[i].length; j++) {
+          const tmpPart = allPartsList[sortedDancerTmp[i]][j];
+          if (tmpPart.type === "FIBER") {
+            if (!allFiberColors.has(status[i][j][0])) {
+              throw new Error(`Fiber color '${status[i][j][0]}' not found!!`);
+            }
+          }
+        }
+      }
+    });
+    console.log("Data valid!!");
```

Comment:
- Make sure every frame's dancers number & parts number are equivalent to real dancers number & parts number.
- Make sure the LED data's shape is correct
  - 2 dimension
  - shape => (None, 4)
- Make sure every required keys in every object exist